### PR TITLE
[core][Darwin] SDK objects should hold weak pointers to the core style objects

### LINKED ISSUE #15333: [ios] Crashes caused by dangling rawSource pointer after new style is parsed.
If a developer retains a `MGLSource`, then sets a new style URL (loading a new style), then that `MGLSource` will end up with a dangling `rawSource` pointer.

In `Style::Impl::parse` we have `sources.clear();` without a mechanism that updates the platform object. The fact this doesn't crash immediately when accessing the rawSource pointer is just pure luck.

Interestingly @asheemmamoowala and I worked on [a similar issue](https://github.com/mapbox/mapbox-gl-native/pull/11553) with `CustomLayer`/`RenderCustomLayer`. I think a similar mechanism may be needed here; it's a generic enough problem.

Related: https://github.com/mapbox/mapbox-gl-native/issues/6180

/cc @samfader @tmpsantos @frederoni

### PR DESCRIPTION
**Pull request by pozdnyakov**
Thus, we can avoid crashes caused by dangling raw{Source|Layer} pointer after new style is parsed.

Fixes #15333

- [x]  Add integration test

### CODE REVIEW

File: platform/darwin/src/MGLSource.mm [hunk 1]
```diff
@@ -33,13 +36,18 @@ - (instancetype)initWithIdentifier:(NSString *)identifier
...
-        _rawSource = rawSource;
-        _rawSource->peer = SourceWrapper { self };
+        _weakSource = rawSource->makeWeakPtr();
+        rawSource->peer = SourceWrapper { self };
         _mapView = mapView;
     }
     return self;
 }
 
+- (mbgl::style::Source *)rawSource
```
Comment (julianrex) [inline platform/darwin/src/MGLSource.mm]: Need to look at any nullability concerns here - 

Though I think it's only the `MGLRasterTileSource` `rawSource` property that might need to be tweaked as nullable - though this is still a private header, so not a huge pain:

https://github.com/mapbox/mapbox-gl-native/blob/d41f6c9f74580a8825da55b9ca63587b0ff1195f/platform/darwin/src/MGLRasterTileSource_Private.h#L11-L15

File: platform/darwin/src/MGLSource_Private.h [hunk 2]
```diff
@@ -18,6 +19,21 @@ struct SourceWrapper {
     __weak MGLSource *source;
 };
 
+/**
+ Assert that the style source is valid.
+ 
+ This macro should be used at the beginning of any public-facing instance method
+ of `MGLSource` and its subclasses. For private methods, an assertion is more appropriate.
+ */
+#define MGLAssertStyleSourceIsValid() \
```
[the remark this answers is not in the document]
Comment (pozdnyakov) [reply platform/darwin/src/MGLSource_Private.h]: Agree :+1: Let's do it in a follow-up PR.

File: platform/darwin/src/MGLStyleLayer_Private.h [hunk 3]
```diff
@@ -27,7 +27,8 @@ struct LayerWrapper {
         if (!self.rawLayer) { \
             [NSException raise:MGLInvalidStyleLayerException \
                         format: \
-            @"-[MGLStyle removeLayer:] has been called " \
+            @"-Either this layer got invalidated after the style change or " \
```
[the remark this answers is not in the document]
Comment (pozdnyakov) [reply platform/darwin/src/MGLStyleLayer_Private.h]: Ah, thanks for the explanation!

**Comment by pozdnyakov** [comment]
@julianrex Thanks for your comments! considered in the latest commit.

File: include/mbgl/style/layer.hpp [hunk 4]
```diff
@@ -6,6 +6,8 @@
 #include <mbgl/style/types.hpp>
 #include <mbgl/style/conversion.hpp>
 
+#include <mapbox/weak.hpp>
+
 #include <cassert>
```
Comment (alexshalamov) [inline include/mbgl/style/layer.hpp]: nit unrelated to this PR: looks like cassert and stdexcept can be removed

**Comment by julianrex** [comment]
@pozdnyakov FYI updated with some tweaks, and an additional test.

**Comment by julianrex** [comment]
@captainbarbosa can you review the ios/macos bits please?

**Pull request merged**